cbmc/proofs/mqtt_state.c

```diff
@@ -0,0 +1,560 @@
+{
+  size_t length;
+  // Assumption avoids arithmetic overflow in malloc, rechecked in valid_*
+  __CPROVER_assume( length < CBMC_MAX_OBJECT_SIZE - sizeof( *pElt ) );
+
+  if ( pElt == NULL ) pElt = malloc_can_fail( sizeof( *pElt ) + length );
+  if ( pElt == NULL ) return NULL;
+
+  pElt->link.pPrevious = NULL;
+  pElt->link.pNext = NULL;
+  pElt->topicFilterLength = length;
+  return pElt;
+}
+
+bool valid_IotMqttSubscriptionListElt( const _mqttSubscription_t *pElt )
+{
+  if ( pElt == NULL ) return false;
+
+  return
+#ifdef TOPIC_LENGTH_MAX
+    pElt->topicFilterLength < TOPIC_LENGTH_MAX &&
+#endif
+    pElt->topicFilterLength < CBMC_MAX_OBJECT_SIZE - sizeof( *pElt ) &&
+    __CPROVER_r_ok( pElt->pTopicFilter, pElt->topicFilterLength ) &&
+    __CPROVER_w_ok( pElt->pTopicFilter, pElt->topicFilterLength );
+}
+
+// Subscription linked lists
+
+IotListDouble_t *allocate_IotMqttSubscriptionList( IotListDouble_t *pSub,
+						   size_t length )
+{
+  if ( pSub == NULL ) pSub = malloc_can_fail( sizeof( *pSub ) );
+  if ( pSub == NULL ) return NULL;
+
+  IotListDouble_Create( pSub );
+  for ( int i = 0; i < length; i++ ) {
+    _mqttSubscription_t *pElt = allocate_IotMqttSubscriptionListElt( NULL );
+    __CPROVER_assume( pElt );
+    IotListDouble_InsertHead( pSub, &( pElt->link ) );
+  }
+  return pSub;
+}
+
+bool valid_IotMqttSubscriptionList( const IotListDouble_t *pSub,
+				    const size_t length )
+{
+  if ( pSub == NULL ) return false;
+
+  bool result = 1;
+  IotListDouble_t *pLink;
+  IotContainers_ForEach( pSub, pLink ) {
+    _mqttSubscription_t
+      *pElt = IotLink_Container( _mqttSubscription_t, pLink, link );
+    result = result && valid_IotMqttSubscriptionListElt( pElt );
+  }
+
+  return
+#ifdef SUBSCRIPTION_COUNT_MAX
+    length < SUBSCRIPTION_COUNT_MAX &&
+#endif
+    result;
+}
+
+/****************************************************************
+ * IotMqttPublishInfo
+ ****************************************************************/
+
+IotMqttPublishInfo_t *allocate_IotMqttPublishInfo( IotMqttPublishInfo_t *pInfo )
+{
+  if ( pInfo == NULL ) pInfo = malloc_can_fail( sizeof( *pInfo ) );
+  if ( pInfo == NULL ) return NULL;
+
+  pInfo->pTopicName = malloc_can_fail( pInfo->topicNameLength );
+  // assumption here is checked below in valid_
+  __CPROVER_assume( VALID_CBMC_SIZE( pInfo->payloadLength ) );
+  pInfo->pPayload = malloc_can_fail( pInfo->payloadLength );
+  return pInfo;
+}
+
+bool valid_IotMqttPublishInfo( const IotMqttPublishInfo_t *pInfo )
+{
+  return
+    pInfo &&
+
+    VALID_STRING( pInfo->pTopicName, pInfo->topicNameLength ) &&
+    VALID_CBMC_SIZE( pInfo->topicNameLength ) &&
+    pInfo->pTopicName != NULL &&
+
+    VALID_STRING( pInfo->pPayload, pInfo->payloadLength ) &&
+    VALID_CBMC_SIZE( pInfo->payloadLength ) &&
+
+    pInfo->retryMs <= IOT_MQTT_RETRY_MS_CEILING &&
+    pInfo->retryMs > 0  &&
+
+    // TODO: experiment with removing these assumptions
+    pInfo->topicNameLength < 0xFFFF &&
```

Comment:
magic number

Reply:
The member topicNameLength is a uint16.  I've added a comment.  I added this because I thought I saw cbmc giving topicNameLength a large value, but I may have been mistaken.  This may indeed be a redundant statement.

Reply:
Why < instead of >=? And why not UINT16_MAX as the variable?

Reply:
Using UINT16_MAX.  Reverting to <= instead of < until something breaks, my memory is that some headroom was required but I can't reconstruct it.